src/java.desktop/windows/classes/com/sun/java/swing/plaf/windows/XPStyle.java

```diff
@@ -675,14 +679,20 @@ protected void paintToImage(Component c, Image image, Graphics g,
             w = bi.getWidth();
             h = bi.getHeight();
 
+            // Get DPI to pass further to ThemeReader.paintBackground()
+            Graphics2D g2d = (Graphics2D) g;
+            AffineTransform  at = g2d.getTransform();
+            int dpi = (int)(at.getScaleX() * 96);
```

Comment:
I think this will not work if the image is not VolatileImage, the scale will be = 1 and as a result, the image could be twice smaller than expected on the HiDPI monitor since rescale operation in another place is deleted. Please double-check.

Reply:
Could you elaborate, please? How can this be not a `VolatileImage`?

If the components are painted into a `BufferedImage`, the scale needs to be specified explicitly if it's the intent.